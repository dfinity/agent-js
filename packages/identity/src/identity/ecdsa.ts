import { DerEncodedPublicKey, PublicKey, Signature, SignIdentity } from '@dfinity/agent';

/**
 * Options used in a {@link ECDSAKeyIdentity}
 */
export type CryptoKeyOptions = {
  extractable?: boolean;
  keyUsages?: KeyUsage[];
  subtleCrypto?: SubtleCrypto;
};

export class CryptoError extends Error {
  constructor(public readonly message: string) {
    super(message);
    Object.setPrototypeOf(this, CryptoError.prototype);
  }
}

export interface DerCryptoKey extends CryptoKey {
  toDer: () => DerEncodedPublicKey;
}

/**
 * Utility method to ensure that a subtleCrypto implementation is provided or is available in the global context
 * @param subtleCrypto SubtleCrypto implementation
 * @returns subleCrypto
 */
function _getEffectiveCrypto(subtleCrypto: CryptoKeyOptions['subtleCrypto']): SubtleCrypto {
  if (subtleCrypto) {
    return subtleCrypto;
  } else if (typeof crypto !== 'undefined' && crypto['subtle']) {
    return crypto.subtle;
  } else {
    throw new CryptoError(
      'Global crypto was not available and none was provided. Please inlcude a SubtleCrypto implementation. See https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto',
    );
  }
}

/**
 * An identity interface that wraps an ECDSA keypair using the P-256 named curve. Supports DER-encoding and decoding for agent calls
 */
export class ECDSAKeyIdentity extends SignIdentity {
  /**
   * Generates a randomly generated identity for use in calls to the Internet Computer.
   * @param {CryptoKeyOptions} options optional settings
   * @param {CryptoKeyOptions['extractable']} options.extractable - whether the key should allow itself to be used. Set to false for maximum security.
   * @param {CryptoKeyOptions['keyUsages']} options.keyUsages - a list of key usages that the key can be used for
   * @param {CryptoKeyOptions['subtleCrypto']} options.subtleCrypto interface
   * @constructs ECDSAKeyIdentity
   * @returns a {@link ECDSAKeyIdentity}
   */
  public static async generate(options?: CryptoKeyOptions): Promise<ECDSAKeyIdentity> {
    const { extractable = false, keyUsages = ['sign', 'verify'], subtleCrypto } = options ?? {};
    const effectiveCrypto = _getEffectiveCrypto(subtleCrypto);
    const keyPair = await effectiveCrypto.generateKey(
      {
        name: 'ECDSA',
        namedCurve: 'P-256',
      },
      extractable,
      keyUsages,
    );
    const derKey = (await effectiveCrypto.exportKey(
      'spki',
      keyPair.publicKey,
    )) as DerEncodedPublicKey;

    return new this(keyPair, derKey, effectiveCrypto);
  }

  /**
   * generates an identity from a public and private key. Please ensure that you are generating these keys securely and protect the user's private key
   * @param keyPair a {@link CryptoKeyPair}
   * @param subtleCrypto a {@link SubtleCrypto} interface in case one is not available globally
   * @returns an {@link ECDSAKeyIdentity}
   */
  public static async fromKeyPair(
    keyPair: CryptoKeyPair | { privateKey: CryptoKey; publicKey: CryptoKey },
    subtleCrypto?: SubtleCrypto,
  ): Promise<ECDSAKeyIdentity> {
    const effectiveCrypto = _getEffectiveCrypto(subtleCrypto);
    const derKey = (await effectiveCrypto.exportKey(
      'spki',
      keyPair.publicKey,
    )) as DerEncodedPublicKey;
    return new ECDSAKeyIdentity(keyPair, derKey, effectiveCrypto);
  }

  protected _derKey: DerEncodedPublicKey;
  protected _keyPair: CryptoKeyPair;
  protected _subtleCrypto: SubtleCrypto;

  // `fromKeyPair` and `generate` should be used for instantiation, not this constructor.
  protected constructor(
    keyPair: CryptoKeyPair,
    derKey: DerEncodedPublicKey,
    subtleCrypto: SubtleCrypto,
  ) {
    super();
    this._keyPair = keyPair;
    this._derKey = derKey;
    this._subtleCrypto = subtleCrypto;
  }

  /**
   * Return the internally-used key pair.
   * @returns a {@link CryptoKeyPair}
   */
  public getKeyPair(): CryptoKeyPair {
    return this._keyPair;
  }

  /**
   * Return the public key.
   * @returns an {@link DerCryptoKey}
   */
  public getPublicKey(): DerCryptoKey {
    const derKey = this._derKey;
    const key: DerCryptoKey = Object.create(this._keyPair.publicKey);
    key.toDer = function () {
      return derKey;
    };

    return key;
  }

  /**
   * Signs a blob of data, with this identity's private key.
   * @param {ArrayBuffer} challenge - challenge to sign with this identity's secretKey, producing a signature
   * @returns {Promise<Signature>} signature
   */
  public async sign(challenge: ArrayBuffer): Promise<Signature> {
    const params: EcdsaParams = {
      name: 'ECDSA',
      hash: { name: 'SHA-256' },
    };
    this._keyPair.privateKey;
    const signature = await this._subtleCrypto.sign(params, this._keyPair.privateKey, challenge);

    return signature as Signature;
  }
}

export default ECDSAKeyIdentity;
