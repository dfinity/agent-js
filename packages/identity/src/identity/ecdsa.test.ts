import { sha256 } from 'js-sha256';
import { ECDSAKeyIdentity, ECDSAPublicKey } from './ecdsa';
import { Crypto } from '@peculiar/webcrypto';

const subtle = new Crypto().subtle;

// Generated by Uint8Array(randomBytes(32)), should not change
const goldenSeed = {
  kty: 'EC',
  crv: 'P-256',
  key_ops: [],
  ext: true,
  x: '6IT6R5WPGgc2tCZhJbLy16TFljSK-HMo5B7YPk4n3-A',
  y: 'TUJE5myib04qYxvJmM-LvYr8Vy1PxguV5eR9GQ9FguU',
};

// const subtle = (crypto.webcrypto as unknown as Crypto).subtle;

describe('ECDSAPublicKey Tests', () => {
  it('can generate a public key', async () => {
    const key = await ECDSAPublicKey.generate({
      subtleCrypto: subtle,
    });

    expect(Object.keys(key)).toMatchSnapshot();
  });
  test('create from an existing public key', async () => {
    const key = await ECDSAPublicKey.fromJWK(goldenSeed, { subtleCrypto: subtle });
    expect(key).toMatchSnapshot();
  });

  test('DER encoding of ECDSA keys', async () => {
    const key = await ECDSAPublicKey.fromJWK(goldenSeed, { subtleCrypto: subtle });
    expect(new Uint8Array(key.toDer())).toMatchSnapshot();
  });

  it('should resolve the same key from a DER-encoded format as from JWK', async () => {
    const key = await ECDSAPublicKey.fromJWK(goldenSeed, {
      extractable: true,
      keyUsages: [],
      subtleCrypto: subtle,
    });

    const encoded = key.toDer();
    expect(new Uint8Array(encoded)).toMatchSnapshot();

    const decodedKey = ECDSAPublicKey.fromDer(encoded, {
      extractable: true,
      keyUsages: [],
      subtleCrypto: subtle,
    });
    expect(key.toJwk()).toEqual((await decodedKey).toJwk());
  });

  test.skip('DER decoding of invalid keys', async () => {
    //     // Too short.
    //     expect(() => {
    //       ECDSAPublicKey.fromDer(
    //         fromHexString(
    //           '3056301006072a8648ce3d020106052b8104000a0342000401ec030acd7d1199f73ae3469329c114944e0693c89502f850bcc6bad397a5956767c79b410c29ac6f587eec84878020fdb54ba002a79b02aa153fe47b6',
    //         ) as DerEncodedPublicKey,
    //       );
    //     }).toThrowError('DER payload mismatch: Expected length 65 actual length 63');
    //     // Too long.
    //     expect(() => {
    //       ECDSAPublicKey.fromDer(
    //         fromHexString(
    //           '3056301006072a8648ce3d020106052b8104000a0342000401ec030acd7d1199f73ae3469329c114944e0693c89502f850bcc6bad397a5956767c79b410c29ac6f587eec84878020fdb54ba002a79b02aa153fe47b6ffd33' +
    //             '1b42211ce',
    //         ) as DerEncodedPublicKey,
    //       );
    //     }).toThrowError('DER payload mismatch: Expected length 65 actual length 70');
    //     // Invalid DER-encoding.
    //     expect(() => {
    //       ECDSAPublicKey.fromDer(
    //         fromHexString(
    //           '0693c89502f850bcc6bad397a5956767c79b410c29ac6f54fdac09ea93a1b9b744b5f19f091ada7978ceb2f045875bca8ef9b75fa8061704e76de023c6a23d77a118c5c8d0f5efaf0dbbfcc3702d5590604717f639f6f00d',
    //         ) as DerEncodedPublicKey,
    //       );
    //     }).toThrowError('Expected: sequence');
  });
});

describe('ECDSAKeyIdentity Tests', () => {
  test('can encode and decode to/from keyPair', async () => {
    const identity = await ECDSAKeyIdentity.generate({ subtleCrypto: subtle });
    const keyPair = identity.getKeyPair();
    const identity2 = await ECDSAKeyIdentity.fromKeyPair(keyPair, subtle);
    expect(new Uint8Array(identity.getPublicKey().toDer())).toEqual(
      new Uint8Array(identity2.getPublicKey().toDer()),
    );
  });

  test('getKeyPair should return a copy of the key pair', async () => {
    const identity = await ECDSAKeyIdentity.generate({ subtleCrypto: subtle });
    const keyPair = identity.getKeyPair();
    expect(keyPair.publicKey).toBeTruthy();
    expect(keyPair.privateKey).toBeTruthy();
  });

  test('message signature is valid', async () => {
    const identity = await ECDSAKeyIdentity.generate({ subtleCrypto: subtle });
    const message = 'Hello world. ECDSA test here';
    const challenge = new TextEncoder().encode(message);
    const signature = await identity.sign(challenge);
    const hash = sha256.create();
    hash.update(challenge);
    const isValid = await subtle.verify(
      {
        name: 'ECDSA',
        hash: { name: 'SHA-256' },
      },
      identity.getKeyPair().publicKey,
      signature,
      challenge,
    );
    expect(isValid).toBe(true);
  });
});
