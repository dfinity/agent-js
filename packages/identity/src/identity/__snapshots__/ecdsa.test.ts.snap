// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ECDSAPublicKey Tests DER encoding of ECDSA keys 1`] = `
Uint8Array [
  48,
  89,
  48,
  19,
  6,
  7,
  42,
  134,
  72,
  206,
  61,
  2,
  1,
  6,
  8,
  42,
  134,
  72,
  206,
  61,
  3,
  1,
  7,
  3,
  66,
  0,
  4,
  232,
  132,
  250,
  71,
  149,
  143,
  26,
  7,
  54,
  180,
  38,
  97,
  37,
  178,
  242,
  215,
  164,
  197,
  150,
  52,
  138,
  248,
  115,
  40,
  228,
  30,
  216,
  62,
  78,
  39,
  223,
  224,
  77,
  66,
  68,
  230,
  108,
  162,
  111,
  78,
  42,
  99,
  27,
  201,
  152,
  207,
  139,
  189,
  138,
  252,
  87,
  45,
  79,
  198,
  11,
  149,
  229,
  228,
  125,
  25,
  15,
  69,
  130,
  229,
]
`;

exports[`ECDSAPublicKey Tests can generate a public key 1`] = `
Array [
  "rawKey",
  "derKey",
  "jwk",
  "algorithm",
  "type",
  "usages",
]
`;

exports[`ECDSAPublicKey Tests create from an existing public key 1`] = `
ECDSAPublicKey {
  "algorithm": Object {
    "name": "ECDSA",
    "namedCurve": "P-256",
  },
  "derKey": ArrayBuffer [],
  "jwk": Object {
    "crv": "P-256",
    "ext": true,
    "key_ops": Array [],
    "kty": "EC",
    "x": "6IT6R5WPGgc2tCZhJbLy16TFljSK-HMo5B7YPk4n3-A",
    "y": "TUJE5myib04qYxvJmM-LvYr8Vy1PxguV5eR9GQ9FguU",
  },
  "rawKey": ArrayBuffer [],
  "type": "public",
  "usages": Array [],
}
`;

exports[`ECDSAPublicKey Tests should resolve the same key from a DER-encoded format as from JWK 1`] = `
Uint8Array [
  48,
  89,
  48,
  19,
  6,
  7,
  42,
  134,
  72,
  206,
  61,
  2,
  1,
  6,
  8,
  42,
  134,
  72,
  206,
  61,
  3,
  1,
  7,
  3,
  66,
  0,
  4,
  232,
  132,
  250,
  71,
  149,
  143,
  26,
  7,
  54,
  180,
  38,
  97,
  37,
  178,
  242,
  215,
  164,
  197,
  150,
  52,
  138,
  248,
  115,
  40,
  228,
  30,
  216,
  62,
  78,
  39,
  223,
  224,
  77,
  66,
  68,
  230,
  108,
  162,
  111,
  78,
  42,
  99,
  27,
  201,
  152,
  207,
  139,
  189,
  138,
  252,
  87,
  45,
  79,
  198,
  11,
  149,
  229,
  228,
  125,
  25,
  15,
  69,
  130,
  229,
]
`;
